[RISCV] Select and/or/xor with certain constants to Zbb ANDN/ORN/XNOR

pfusik opened the pull request:
(and X, (C<<12|0xfff)) -> (ANDN X, ~C<<12)
    (or  X, (C<<12|0xfff)) -> (ORN  X, ~C<<12)
    (xor X, (C<<12|0xfff)) -> (XNOR X, ~C<<12)

Emits better code, typically by avoiding an `ADDI HI, -1` instruction.

Co-authored-by: Craig Topper <craig.topper@sifive.com>

pfusik commented:
I'd like to implement an optimization so that

```c
long orlow(long x) {
    return x | ((1L << 24) - 1);
}
```

that is now compiled into:

```
        lui     a1, 4096
        addiw   a1, a1, -1
        or      a0, a0, a1
        ret
```

gets instead compiled to:

```
        lui     a1, -4096
        orn     a0, a0, a1
        ret
```

and similarly for AND and XOR.

I'm trying to implement this with tablegen patterns, but I'm new to that technology and I got stuck with the following error:

```
[16/130 -j3] Building RISCVGenGlobalISel.inc...
FAILED: lib/Target/RISCV/RISCVGenGlobalISel.inc /home/p.fusik/upstream/llvm-project/build/lib/Target/RISCV/RISCVGenGlobalISel.inc
cd /home/p.fusik/upstream/llvm-project/build && /home/p.fusik/upstream/llvm-project/build/bin/llvm-tblgen -gen-global-isel -I /home/p.fusik/upstream/llvm-project/llvm/lib/Target/RISCV -I/home/p.fusik/upstream/llvm-project/build/include -I/home/p.fusik/upstream/llvm-project/llvm/include -I /home/p.fusik/upstream/llvm-project/llvm/lib/Target /home/p.fusik/upstream/llvm-project/llvm/lib/Target/RISCV/RISCVGISel.td --write-if-changed -o lib/Target/RISCV/RISCVGenGlobalISel.inc -d lib/Target/RISCV/RISCVGenGlobalISel.inc.d
anonymous_10632:        (or:{ *:[i32] m1:[i64] } (vt:{} GPR:{ *:[i16 i32] m1:[i16 i32 i64 f64] }:$rs1), (imm:{ *:[i32] m1:[i64] })<<P:Predicate_simm32fff>>:$imm)
Included from /home/p.fusik/upstream/llvm-project/llvm/lib/Target/RISCV/RISCVGISel.td:16:
Included from /home/p.fusik/upstream/llvm-project/llvm/lib/Target/RISCV/RISCV.td:36:
Included from /home/p.fusik/upstream/llvm-project/llvm/lib/Target/RISCV/RISCVInstrInfo.td:2113:
/home/p.fusik/upstream/llvm-project/llvm/lib/Target/RISCV/RISCVInstrInfoZb.td:499:1: error: In anonymous_10632: vt node requires exactly 0 operands!
def : PatGprSimm32fff<or, ORN>;
^
anonymous_10632:        (or:{ *:[i32] m1:[i64] } (vt:{} GPR:{ *:[i16 i32] m1:[i16 i32 i64 f64] }:$rs1), (imm:{ *:[i32] m1:[i64] })<<P:Predicate_simm32fff>>:$imm)
anonymous_10632:        (ORN:{ *:[i32] m1:[i64] } GPR:{ *:[i16 i32] m1:[i16 i32 i64 f64] }:$rs1, (NotImm:{ *:[i16 i32] m1:[i16 i32 i64 f64] } (imm:{ *:[i32] m1:[i64] })<<P:Predicate_simm32fff>>:$imm))
Included from /home/p.fusik/upstream/llvm-project/llvm/lib/Target/RISCV/RISCVGISel.td:16:
Included from /home/p.fusik/upstream/llvm-project/llvm/lib/Target/RISCV/RISCV.td:36:
Included from /home/p.fusik/upstream/llvm-project/llvm/lib/Target/RISCV/RISCVInstrInfo.td:2113:
/home/p.fusik/upstream/llvm-project/llvm/lib/Target/RISCV/RISCVInstrInfoZb.td:499:1: error: In anonymous_10632: Could not infer all types in pattern result!
def : PatGprSimm32fff<or, ORN>;
^
```

Please let me know how can I fix it or if perhaps this approach is completely wrong.

mshockwave commented on llvm/lib/Target/RISCV/RISCVInstrInfoZb.td:
```diff
@@ -481,6 +481,25 @@ def : Pat<(XLenVT (or  GPR:$rs1, (not GPR:$rs2))), (ORN  GPR:$rs1, GPR:$rs2)>;
+}]>;
+
+class PatGprSimm32fff<SDPatternOperator OpNode, RVInst Inst>
+    : Pat<(XLenVT (OpNode (vt GPR:$rs1), simm32fff:$imm)),
```
`vt` by itself is a SDNode representing ValueType (https://github.com/llvm/llvm-project/blob/1c16807d0dd740ace0f21ed29d1381a0078f745e/llvm/include/llvm/Target/TargetSelectionDAG.td#L367). For example, `sext_inreg` has a vt operand attached at the end to specify the original value type. Therefore you shouldn't put it here and that's what the error message told you.

You might see lots of `vt` popping up here and there in other parts of the backend, and that's because we (for better or for worse) overloading this word for variable names as well. In which case it's just a variable for plugging in different value types.

I think what you want here might be `(XLenVT GPR:$rs1)`?

topperc replied:
`GPR:$rs1` instead of `(XLenVT GPR:$rs1)` should be find since `XLenVT` is already on the outside of `OpNode`.

topperc commented on llvm/lib/Target/RISCV/RISCVInstrInfoZb.td:
```diff
@@ -481,6 +481,25 @@ def : Pat<(XLenVT (or  GPR:$rs1, (not GPR:$rs2))), (ORN  GPR:$rs1, GPR:$rs2)>;
+
+class PatGprSimm32fff<SDPatternOperator OpNode, RVInst Inst>
+    : Pat<(XLenVT (OpNode (vt GPR:$rs1), simm32fff:$imm)),
+          (Inst GPR:$rs1, (NotImm simm32fff:$imm))>;
```
We need to explicitly create an LUI for the immediate here. And we need to put the immediate into LUI format which requires shifting right by 12 and masking to 20 bits.

topperc replied:
MachineLICM pass should be able to move it outside loops. SelectionDAG will CSE it if its used multiple times in the same basic block. Otherwise MachineCSE pass will be responsible for CSE.

topperc commented:
Here's another case that might be useful to optimize https://godbolt.org/z/MEzP15sas it already generates a `not` but we don't fold it.

pfusik commented:
Force pushed because of a precommit test.

pfusik commented on llvm/test/CodeGen/RISCV/pr84653_pr85190.ll:
```diff
@@ -21,8 +21,7 @@ define i1 @pr84653(i32 %x) {
 ; CHECK-ZBB:       # %bb.0:
 ; CHECK-ZBB-NEXT:    sext.w a1, a0
 ; CHECK-ZBB-NEXT:    lui a2, 524288
```
`lui` is unchanged here because it's a `X ^ ((1 << 31) - 1) -> X ^ ~(1 << 31)` case.

pfusik commented on llvm/test/CodeGen/RISCV/zbb-logic-neg-imm.ll:
```diff
@@ -5,109 +5,77 @@
 ; RV32-LABEL: oraddlow16:
 ; RV32:       # %bb.0:
+; RV32-NEXT:    lui a1, 1048560
+; RV32-NEXT:    orn a0, a0, a1
```
In this case we pessimize. I expect this to be rare. Do we want to address it? How?

lenary replied:
Can you do a `hasOneUse` check on the ConstantSDNode? I don't know whether it's worth trying.

pfusik replied:
Yes, this prevents pessimizations in `oraddlow16` and `addorlow16`. But also disables optimization in `orarray3`.

lenary replied:
I did think about this, but I'm not sure whether we want to be looking at all the users like that during selection, or whether this should be a target-specific DAGCombine rather than a set of selection patterns. I'll defer to craig on that.

topperc replied:
I think I prefer checking the users during selection. If we do a target specific combine we need new ISD nodes to represent the optimization.

pfusik commented on llvm/test/CodeGen/RISCV/zbb-logic-neg-imm.ll:
```diff
@@ -121,15 +89,17 @@ define i32 @addorlow16(i32 %x) {
 ; RV32-NEXT:    add a0, a0, a1
-; RV32-NEXT:    or a0, a0, a1
+; RV32-NEXT:    lui a1, 1048560
+; RV32-NEXT:    orn a0, a0, a1
```
Ditto.

pfusik commented on llvm/test/CodeGen/RISCV/zbb-logic-neg-imm.ll:
```diff
@@ -273,37 +240,21 @@ define i32 @compl(i32 %x) {
 }
 
 define i32 @orlow12(i32 %x) {
```
With Zbs this is:

        ori     a0, a0, 2047
        bseti   a0, a0, 11

and is not affected by this change. Shall I add `+zbs` or test both with and without?

lenary replied:
I would add RUN lines with `+zbs`, rather than adding `+zbs` to the existing RUN lines.

pfusik replied:
Done

pfusik commented on llvm/test/CodeGen/RISCV/zbb-logic-neg-imm.ll:
```diff
@@ -273,37 +240,21 @@ define i32 @compl(i32 %x) {
   ret i32 %or
 }
 
 define i32 @xorlow12(i32 %x) {
```
Similar to the above.

        xori    a0, a0, 2047
        binvi   a0, a0, 11

pfusik commented:
> Here's a rough sketch using a ComplexPattern instead of PatLeaf/SDNodeXForm.

This works great!

I didn't even have to special-case `and` for `sll+srl` etc. But I don't know why. Does the pattern order matter or is there some scoring mechanism?

pfusik commented:
> Here's another case that might be useful to optimize https://godbolt.org/z/MEzP15sas it already generates a `not` but we don't fold it.

I can work on it.

topperc commented on llvm/lib/Target/RISCV/RISCVISelDAGToDAG.cpp:
```diff
@@ -3236,6 +3236,22 @@ bool RISCVDAGToDAGISel::selectSHXADD_UWOp(SDValue N, unsigned ShAmt,
+    return false;
+
+  int64_t Imm = cast<ConstantSDNode>(N)->getSExtValue();
+  if (!(isInt<32>(Imm) && (Imm & 0xfff) == 0xfff && Imm != -1))
```
Push the `!` into the individual conditions and use `||`.

pfusik replied:
Done

pfusik commented on llvm/test/CodeGen/RISCV/zbb-logic-neg-imm.ll:
```diff
@@ -0,0 +1,345 @@
+; RV64-NEXT:    lui a1, 65281
+; RV64-NEXT:    slli a1, a1, 4
+; RV64-NEXT:    addi a1, a1, -1
+; RV64-NEXT:    and a0, a0, a1
```
Here it's possible to optimize a non-int32. Getting the condition right in general seems tricky. Can I call `RISCVMatInt` for `Imm` and `~Imm` to see which one is shorter?

lenary replied:
This wouldn't be the only place to compare two `getIntMatCost`s, so it should be fine, as long as you check the simpler exclusionary conditions first - maybe even including your `isBitwiseLogicOp` loop.

pfusik reviewed (commented):
Added Zbs testing. Added `andimm64` test. Squashed the change so that the effect on tests is clear.

pfusik committed:
[RISCV] Select and/or/xor with certain constants to Zbb ANDN/ORN/XNOR

    (and X, (C<<12|0xfff)) -> (ANDN X, ~C<<12)
    (or  X, (C<<12|0xfff)) -> (ORN  X, ~C<<12)
    (xor X, (C<<12|0xfff)) -> (XNOR X, ~C<<12)

Emits better code, typically by avoiding an `ADDI HI, -1` instruction.

Co-authored-by: Craig Topper <craig.topper@sifive.com>

pfusik commented:
> Here's another case that might be useful to optimize https://godbolt.org/z/MEzP15sas it already generates a `not` but we don't fold it.

Done. Covered by the `andimm64srli` test.

lenary reviewed (approved):
I'm happy with this.

By default, LLVM will squash-merge, so can you pre-commit the new test file before landing this?

pfusik commented:
> By default, LLVM will squash-merge, so can you pre-commit the new test file before landing this?

Do you mean create a new PR for the pre-commit test?

topperc reviewed (approved):
LGTM

topperc commented on llvm/lib/Target/RISCV/RISCVISelDAGToDAG.cpp:
```diff
@@ -3236,6 +3236,35 @@ bool RISCVDAGToDAGISel::selectSHXADD_UWOp(SDValue N, unsigned ShAmt,
+  if ((Imm & 0xfff) != 0xfff || Imm == -1)
+    return false;
+
+  for (const SDNode *U : N->uses()) {
```
This needs to change `N->users()` now.

topperc commented:
> > By default, LLVM will squash-merge, so can you pre-commit the new test file before landing this?
> 
> Do you mean create a new PR for the pre-commit test?

If you have commit access you can directly push it without going through github.

lenary commented:
Nice work!

pfusik commented:
Thank you.
